File: api/handlers/fuels.go
package handlers

import (
	"net/http"
	"time"

	"fleet-management-api/config"
	"fleet-management-api/models"

	"github.com/gin-gonic/gin"
)

// GET /api/fuels
func GetFuels(c *gin.Context) {
	rows, err := config.DB.Query("SELECT id, name, type, price, active_from, created_at FROM fuels ORDER BY name, active_from DESC")
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"success": false, "message": "Server error"})
		return
	}
	defer rows.Close()

	fuels := []models.Fuel{}
	for rows.Next() {
		var f models.Fuel
		var activeFrom time.Time
		err := rows.Scan(&f.ID, &f.Name, &f.Type, &f.Price, &activeFrom, &f.CreatedAt)
		if err == nil {
			f.ActiveFrom = activeFrom.Format("2006-01-02")
			fuels = append(fuels, f)
		}
	}

	c.JSON(http.StatusOK, gin.H{"success": true, "data": fuels})
}

// GET /api/fuels/active
func GetActiveFuels(c *gin.Context) {
	query := `
		SELECT f1.id, f1.name, f1.type, f1.price, f1.active_from, f1.created_at 
		FROM fuels f1
		JOIN (
			SELECT name, MAX(active_from) as max_date 
			FROM fuels 
			WHERE active_from <= CURDATE() 
			GROUP BY name
		) f2 ON f1.name = f2.name AND f1.active_from = f2.max_date
		ORDER BY f1.name`

	rows, err := config.DB.Query(query)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"success": false, "message": "Server error"})
		return
	}
	defer rows.Close()

	fuels := []models.Fuel{}
	for rows.Next() {
		var f models.Fuel
		var activeFrom time.Time
		err := rows.Scan(&f.ID, &f.Name, &f.Type, &f.Price, &activeFrom, &f.CreatedAt)
		if err == nil {
			f.ActiveFrom = activeFrom.Format("2006-01-02")
			fuels = append(fuels, f)
		}
	}

	c.JSON(http.StatusOK, gin.H{"success": true, "data": fuels})
}

// POST /api/fuels
func CreateFuel(c *gin.Context) {
	var input struct {
		Name       string  `json:"name" binding:"required"`
		Type       string  `json:"type" binding:"required"`
		Price      float64 `json:"price" binding:"required"`
		ActiveFrom string  `json:"active_from" binding:"required"`
	}

	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"success": false, "message": "Semua field wajib diisi: " + err.Error()})
		return
	}

	activeDate, err := time.Parse("2006-01-02", input.ActiveFrom)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"success": false, "message": "Format tanggal active_from tidak valid (harus YYYY-MM-DD)"})
		return
	}

	res, err := config.DB.Exec(
		"INSERT INTO fuels (name, type, price, active_from) VALUES (?, ?, ?, ?)",
		input.Name, input.Type, input.Price, activeDate,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"success": false, "message": "Server error"})
		return
	}

	insertID, _ := res.LastInsertId()
	c.JSON(http.StatusCreated, gin.H{"success": true, "data": gin.H{"id": insertID}, "message": "Bahan Bakar berhasil ditambahkan"})
}

// PUT /api/fuels/:id
func UpdateFuel(c *gin.Context) {
	fuelID := c.Param("id")

	var input struct {
		Name       string  `json:"name" binding:"required"`
		Type       string  `json:"type" binding:"required"`
		Price      float64 `json:"price" binding:"required"`
		ActiveFrom string  `json:"active_from" binding:"required"`
	}

	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"success": false, "message": "Semua field wajib diisi: " + err.Error()})
		return
	}

	activeDate, err := time.Parse("2006-01-02", input.ActiveFrom)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"success": false, "message": "Format tanggal active_from tidak valid (harus YYYY-MM-DD)"})
		return
	}

	_, err = config.DB.Exec(
		"UPDATE fuels SET name = ?, type = ?, price = ?, active_from = ? WHERE id = ?",
		input.Name, input.Type, input.Price, activeDate, fuelID,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"success": false, "message": "Server error"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true, "message": "Bahan Bakar berhasil diupdate"})
}

// DELETE /api/fuels/:id
func DeleteFuel(c *gin.Context) {
	fuelID := c.Param("id")

	_, err := config.DB.Exec("DELETE FROM fuels WHERE id = ?", fuelID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"success": false, "message": "Server error"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true, "message": "Bahan Bakar berhasil dihapus"})
}

File: api-node/routes/fuels.js
const express = require('express');
const router = express.Router();
const db = require('../config/db');
const { auth, checkPermission } = require('../middleware/auth');

// GET all fuels (both active & historical)
router.get('/', auth, async (req, res) => {
  try {
    const [rows] = await db.query('SELECT * FROM fuels ORDER BY name, active_from DESC');
    res.json({ success: true, data: rows });
  } catch (error) {
    console.error('Error fetching fuels:', error);
    res.status(500).json({ success: false, message: 'Server error' });
  }
});

// GET active fuels (current active price for each unique fuel name)
router.get('/active', auth, async (req, res) => {
  try {
    const [rows] = await db.query(`
      SELECT f1.* FROM fuels f1
      JOIN (
        SELECT name, MAX(active_from) as max_date 
        FROM fuels 
        WHERE active_from <= CURDATE() 
        GROUP BY name
      ) f2 ON f1.name = f2.name AND f1.active_from = f2.max_date
      ORDER BY f1.name
    `);
    res.json({ success: true, data: rows });
  } catch (error) {
    console.error('Error fetching active fuels:', error);
    res.status(500).json({ success: false, message: 'Server error' });
  }
});

// POST add new fuel price
router.post('/', auth, checkPermission('vehicles.create'), async (req, res) => {
  try {
    const { name, type, price, active_from } = req.body;
    if (!name || !type || !price || !active_from) {
      return res.status(400).json({ success: false, message: 'Semua field wajib diisi' });
    }

    const [result] = await db.query(
      'INSERT INTO fuels (name, type, price, active_from) VALUES (?, ?, ?, ?)',
      [name, type, Number(price), active_from]
    );

    res.status(201).json({ success: true, data: { id: result.insertId }, message: 'Bahan Bakar berhasil ditambahkan' });
  } catch (error) {
    console.error('Error creating fuel:', error);
    res.status(500).json({ success: false, message: 'Server error' });
  }
});

// PUT update fuel price
router.put('/:id', auth, checkPermission('vehicles.edit'), async (req, res) => {
  try {
    const { name, type, price, active_from } = req.body;
    if (!name || !type || !price || !active_from) {
      return res.status(400).json({ success: false, message: 'Semua field wajib diisi' });
    }

    await db.query(
      'UPDATE fuels SET name = ?, type = ?, price = ?, active_from = ? WHERE id = ?',
      [name, type, Number(price), active_from, req.params.id]
    );

    res.json({ success: true, message: 'Bahan Bakar berhasil diupdate' });
  } catch (error) {
    console.error('Error updating fuel:', error);
    res.status(500).json({ success: false, message: 'Server error' });
  }
});

// DELETE fuel price entry
router.delete('/:id', auth, checkPermission('vehicles.edit'), async (req, res) => {
  try {
    await db.query('DELETE FROM fuels WHERE id = ?', [req.params.id]);
    res.json({ success: true, message: 'Bahan Bakar berhasil dihapus' });
  } catch (error) {
    console.error('Error deleting fuel:', error);
    res.status(500).json({ success: false, message: 'Server error' });
  }
});

module.exports = router;
